File: ui-tests/pages/ArticlePage.ts
import { Page, expect } from '@playwright/test';

export class ArticlePage {
  readonly page: Page;

  constructor(page: Page) {
    this.page = page;
  }

  async navigateToEditor() {
    await this.page.goto('/');
    await this.page.click('a.nav-link:has-text("New Post")');
    await this.page.waitForURL('**/editor');
  }

  async createArticle(title: string, description: string, body: string, tags: string[]) {
    await this.page.fill('input[placeholder="Article Title"]', title);
    await this.page.fill('input[placeholder*="this article about?"]', description);
    await this.page.fill('textarea[placeholder*="Write your article"]', body);
    for (const tag of tags) {
      await this.page.fill('input[placeholder="Enter tags"]', tag);
      await this.page.keyboard.press('Enter');
    }
    await this.page.click('button:has-text("Publish Article")');
    await this.page.waitForURL('**/article/**');
  }

  async getArticleTitle(): Promise<string> {
    return this.page.locator('h1').innerText();
  }

  async openOwnArticleFromHome(title: string) {
    await this.page.goto('/');
    await this.page.click(`a:has-text("${title}")`);
    await this.page.waitForURL('**/article/**');
  }

  async clickEditButton() {
    await this.page.click('a:has-text("Edit Article")');
    await this.page.waitForURL('**/editor/**');
  }

    async editArticle(title: string, description: string, body: string) {
    await this.page.fill('input[placeholder="Article Title"]', title);
    await this.page.fill('input[placeholder*="this article about?"]', description);
    await this.page.fill('textarea[placeholder*="Write your article"]', body);
    await Promise.all([
      this.page.waitForURL('**/article/**', { timeout: 10000 }),
      this.page.click('button:has-text("Publish Article")')
    ]);
  }


async openOwnArticleFallback(title: string) {
  await this.page.goto('/');

  // Try to find the article on the homepage (Your Feed or Global Feed)
  const articleOnHome = this.page.locator(`a:has-text("${title}")`).first();
  const isVisible = await articleOnHome.isVisible({ timeout: 3000 }).catch(() => false);

  if (isVisible) {
    await expect(articleOnHome).toBeVisible();
    await articleOnHome.click();
    await this.page.waitForURL('**/article/**');
    return;
  }

  // If not found, navigate to profile and try there
  await this.page.click('a.nav-link[href*="/@"]');
  await this.page.waitForURL('**/@**');

  const articleInProfile = this.page.locator(`a:has-text("${title}")`).first();
  await expect(articleInProfile).toBeVisible();
  await articleInProfile.click();
  await this.page.waitForURL('**/article/**');
}


  async deleteArticle() {
    await this.page.click('button:has-text("Delete Article")');
    await this.page.waitForURL('**/');
  }


  
}

File: mobile-tests/pages/article_page.py
# pages/article_page.py

from appium.webdriver.common.appiumby import AppiumBy
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC

class ArticlePage:
    def __init__(self, driver):
        self.driver = driver

    def navigate_to_editor(self, base_url):
        self.driver.get(base_url)
        self.driver.find_element(AppiumBy.XPATH, '//a[normalize-space()="New Post"]').click()

    def create_article(self, title, description, body, tags):
        self.driver.find_element(AppiumBy.XPATH, '//input[@placeholder="Article Title"]').send_keys(title)
        self.driver.find_element(AppiumBy.XPATH, '//input[contains(@placeholder,"this article about?")]').send_keys(description)
        self.driver.find_element(AppiumBy.XPATH, '//textarea[contains(@placeholder,"Write your article")]').send_keys(body)
        for tag in tags:
            tag_input = self.driver.find_element(AppiumBy.XPATH, '//input[@placeholder="Enter tags"]')
            tag_input.send_keys(tag)
            tag_input.send_keys('\n')  # simulate Enter
        self.driver.find_element(AppiumBy.XPATH, '//button[normalize-space()="Publish Article"]').click()

    def get_article_title(self):
        return self.driver.find_element(AppiumBy.TAG_NAME, 'h1').text

    def open_own_article_from_home(self, base_url, title):
        self.driver.get(base_url)
        self.driver.find_element(AppiumBy.XPATH, f'//a[normalize-space()="{title}"]').click()

    def click_edit_button(self):
        self.driver.find_element(AppiumBy.XPATH, '//a[normalize-space()="Edit Article"]').click()

    def edit_article(self, title, description, body):
        self.driver.find_element(AppiumBy.XPATH, '//input[@placeholder="Article Title"]').clear()
        self.driver.find_element(AppiumBy.XPATH, '//input[@placeholder="Article Title"]').send_keys(title)

        self.driver.find_element(AppiumBy.XPATH, '//input[contains(@placeholder,"this article about?")]').clear()
        self.driver.find_element(AppiumBy.XPATH, '//input[contains(@placeholder,"this article about?")]').send_keys(description)

        self.driver.find_element(AppiumBy.XPATH, '//textarea[contains(@placeholder,"Write your article")]').clear()
        self.driver.find_element(AppiumBy.XPATH, '//textarea[contains(@placeholder,"Write your article")]').send_keys(body)

        self.driver.find_element(AppiumBy.XPATH, '//button[normalize-space()="Publish Article"]').click()

    def open_own_article_fallback(self, base_url, title):
        self.driver.get(base_url)
        try:
            element = WebDriverWait(self.driver, 3).until(
                EC.presence_of_element_located((AppiumBy.XPATH, f'//a[normalize-space()="{title}"]'))
            )
            element.click()
        except:
            # go to profile if not on homepage
            self.driver.find_element(AppiumBy.XPATH, '//a[contains(@href, "/@")]').click()
            WebDriverWait(self.driver, 10).until(
                EC.presence_of_element_located((AppiumBy.XPATH, f'//a[normalize-space()="{title}"]'))
            ).click()

    def delete_article(self):
        self.driver.find_element(AppiumBy.XPATH, '//button[normalize-space()="Delete Article"]').click()
